driver/generic/st77xx.py

```diff
@@ -0,0 +1,405 @@
+            # front/back porch setting in normal/idle/partial modes; 3rd byte (PSEN) 0x00 = disabled
+            (ST7789_PORCTRL, b"\x0C\x0C\x00\x33\x33"),
+            # VGH=14.06V, VGL=-8.87V [Adafruit: 0x14]
+            (ST7789_GCTRL, b"\x35"),
+            # [Adafruit: missing]
+            (ST77XX_DISPOFF, b"\x28"),
+            # power control [Adafruit: 0x2c]
+            (ST7789_LCMCTRL, b"\x3C"),
+            # power control (set VDV and VRD by register write), write VRH and VDV
+            (ST7789_VDVVRHEN, b"\x01"),(ST7789_VRHS, b"\x0B"),(ST7789_VDVS, b"\x20"),
+            # frame rate 60Hz
+            (ST7789_FRCTRL2, b"\x0F"),
+            # power control: AVDD=6.6V, AVCL=-4.8V, VDS=2.4V
+            (ST7789_PWCTRL1, b"\xA4\xA1"),
+            # positive voltage gamma control
+            (ST7789_PVGAMCTRL, b"\xD0\x01\x08\x0F\x11\x2A\x36\x55\x44\x3A\x0B\x06\x11\x20"),
+            # negative voltage gamma control
+            (ST7789_NVGAMCTRL, b"\xD0\x02\x07\x0A\x0B\x18\x34\x43\x4A\x2B\x1B\x1C\x22\x1F"),
+            # content adaptive brightness control and color enhancement: color enhancement on, high enhancement
+            (ST7789_WRCACE, bytes([0b1011_0000])),
+            # display on
+            (ST77XX_DISPON, None, 100),
+            ## FIXME: needs out of sleep mode AGAIN, otherwise will stay bleck the first time on?
+            (ST77XX_SLPOUT, None, 100),
+        ]
+        self._run_seq(init7789)
+        self.apply_rotation(self.rot)
+
+class St77xx_lvgl(object):
+    '''LVGL wrapper for St77xx, not to be instantiated directly.
+
+    * creates and registers LVGL display driver;
+    * allocates buffers (double-buffered by default);
+    * sets the driver callback to the disp_drv_flush_cb method.
+
+    '''
+    def disp_drv_flush_cb(self,disp_drv,area,color):
+        # print(f"({area.x1},{area.y1}..{area.x2},{area.y2})")
+        self._rp2_wait_dma() # wait if not yet done and DMA is being used
+        # blit in background
+        self.blit(area.x1,area.y1,w:=(area.x2-area.x1+1),h:=(area.y2-area.y1+1),disp_drv.draw_buf.buf_act.__dereference__(2*w*h),is_blocking=False)
+        self.disp_drv.flush_ready()
+    def __init__(self,doublebuffer=True):
+        import lvgl as lv
+        import lv_utils
+
+        if lv.COLOR.DEPTH!=16 or not lv.COLOR_16.SWAP: raise RuntimeError(f'LVGL *must* be compiled with 16bit color depth and swapped bytes (current: lv.COLOR.DEPTH={lv.COLOR.DEPTH}, lv.COLOR_16.SWAP={lv.COLOR_16.SWAP})')
+
+        if not lv.is_initialized(): lv.init()
+        # create event loop if not yet present
+        if not lv_utils.event_loop.is_running(): self.event_loop=lv_utils.event_loop()
+
+        # attach all to self to avoid objects' refcount dropping to zero when the scope is exited
+        self.disp_draw_buf=lv.disp_draw_buf_t()
+        self.disp_draw_buf.init(fb1:=bytearray(self.width*2*32),bytearray(self.width*2*32) if doublebuffer else None,len(fb1)//lv.color_t.__SIZE__)
```

Comment:
Could you explain the `*2*32` multiplication?  How did you decide on this buffer size?  
Why not let the user control the buffer size? On ili9XXX we use `factor` to allow the user control how much of the display buffer should be allocated. (factor 4 means allocate only 1/4 of the display size, and update the display in 8 parts).

Reply:
Added *factor*, thanks for the suggestion (`2*32` rows was 32 rows with 2 bytes per pixel).